scripts/rest_service.py

```diff
@@ -0,0 +1,303 @@
+
+    def add_job(self, job: psij.Job) -> None:
+
+        self._jobs[job.id] = job
+
+    def get_job(self, jobid: str) -> Optional[psij.Job]:
+
+        return self._jobs.get(jobid)
+
+    def list_jobs(self) -> List[str]:
+
+        return list(self._jobs.keys())
+
+    def get_msg(self) -> Any:
+        try:
+            return self._queue.get_nowait()
+        except Exception:
+            pass
+
+    def send(self, msg: Any) -> None:
+        self._queue.put(msg)
+
+
+# ------------------------------------------------------------------------------
+#
+class Service(object):
+
+    def __init__(self, app: FastAPI) -> None:
+
+        self._clients: Dict[str, _Client] = dict()
+        self._deserialize = psij.Import()
+        self._log = logging.getLogger('psij')
+        self._cnt: int = 0
+
+        # ----------------------------------------------------------------------
+        # websocket endpoint at which client can regoister for state updates
+        @app.websocket("/ws/{cid}")
+        async def ws_endpoint(ws: WebSocket, cid: str) -> None:
+
+            await ws.accept()
+
+            client = self._clients.get(cid)
+            if not client:
+                self._log.error("refuse ws for %s" % cid)
+                raise ValueError("unknown client cid %s" % cid)
+
+            self._log.info("accept ws for %s" % cid)
+            try:
+                # keep this async task alive as long as there are messages to
+                # send and the websocket is alive
+                while True:
+                    msg = client.get_msg()
+                    if msg:
+                        await ws.send_json(msg)
+                    else:
+                        await asyncio.sleep(0.1)
+
+            except WebSocketDisconnect:
+                self._log.info("dropped ws for %s" % cid)
+        # ----------------------------------------------------------------------
+
+    def _status_callback(self, cid: str, job: psij.Job, status: psij.JobStatus) -> None:
+
+        client = self._clients.get(cid)
+        if not client:
+            print("unknown client cid %s" % cid)
+            return
+
+        msg = {'jobid': job.id,
+               'time': status.time,
+               'message': status.message,
+               'state': str(status.state),
+               'metadata': status.metadata,
+               'exit_code': status.exit_code}
+        self._log.debug('cb: %s: %s', cid, msg)
+        client.send(msg)
+
+    def _request_register(self, name: str, url: Optional[str] = None) -> str:
+        '''
+        parameters:
+            name:str: name of psij executor to use
+            url:str: optional URL to be passed to backend executor
+
+        returns:
+            str: unique ID identifying the registered client
+        '''
+
+        # register new client
+        cid = 'client.%04d' % self._cnt
+        self._cnt += 1
+
+        # create executor
+        jex = psij.JobExecutor.get_instance(name=name, url=url)
```

Comment:
This works for now. What we might want to consider is different executor configs. The other thing that might be relevant is that either `JobExecutor.get_instance` guarantees referential transparency, or we cache executor instances here.

Reply:
Ack